Teleport Plugins Documentation

benarent opened the pull request:
This PR ports over Install docs from the Teleport-Plugin Repository. 

Todo
- [x] Add / port over screenshots
- [x] Test Setup 
- [ ] Record Setup Videos

benarent committed:
Base fork for 4.3 docs

benarent committed:
wip - Plugins

benarent committed:
Remove alpha and transfer location.

benarent committed:
Ported over setup from Plugins

benarent commented:
@stevenGravy Thanks for the recent language changes, can you please make any other change directly to this PR.

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_jira_server.md:
````diff
@@ -0,0 +1,172 @@
+# https-cert-file = "/var/lib/teleport/plugins/jirabot/server.crt" # TLS certificate
+
+[log]
+output = "stderr" # Logger output. Could be "stdout", "stderr" or "/var/lib/teleport/jirabot.log"
+severity = "INFO" # Logger severity. Could be "INFO", "ERROR", "DEBUG" or "WARN"`
+
+```
+
+The `[teleport]` section describes where is the teleport service running, and what keys should the plugin use to authenticate itself. Use the keys that you've generated [above in exporting your Certificate section](#Export access-plugin Certificate).
+
+The `[jira]` section requires a few things: 
+1. Your Jira Cloud or Jira Server URL. For Jira Cloud, it looks something like yourcompany.atlassian.net. 
+2. Your username on Jira, i.e. ben@gravitational.com
+3. Your Jira API token that you've created above. 
+4. And the Jira Project key, available in Project settings. 
+
+`[http]` setting block describes how the Plugin's HTTP server works. The HTTP server is responsible for listening for updates from Jira, and processing updates, like when someone drags a task from Inbox to Approved column. 
+
+You must provide an address the server should listen on, and a certificate to use, unless you plan on running with `--insecure-no-tls`, which we don't recommend in production. 
+
+
+## Testing
+
+You should be able to run the Teleport plugin now! 
+
+```bash
+teleport-jirabot start
+```
+
+The log output should look familiar to what Teleport service logs. You should see that it connected to Teleport, and is listening for new Teleport requests and Jira webhooks. 
+
+Go ahead and test it: 
+
+```bash
+tsh login --request-roles=admin
+```
+
+That should create a new permission request on Teleport (you can test if it did with `tctl request ls` ), and you should see a new task on your Jira project board.
+
+## FAQ
````
Same corrections to this as to the JIRA Cloud page.

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_mattermost.md:
```diff
@@ -0,0 +1,152 @@
+The new bot account will need Post All permission. 
+
+The confirmation screen after you've created the bot will give you the access token.
+ We'll use this in the config later.
```
Space at beginning of line

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_mattermost.md:
```diff
@@ -0,0 +1,152 @@
+
+#### Create an access-plugin role and user within Teleport 
+First off, using an existing Teleport Cluster, we are going to create a new Teleport
+ User and Role to access Teleport.
```
Space at beginning of line

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_mattermost.md:
```diff
@@ -0,0 +1,152 @@
+
+#### Create User and Role for access. 
+Log into Teleport Authentication Server, this is where you normally run `tctl`. Don't
+ change the username and the role name, it should be `access-plugin` for the plugin 
+ to work correctly.
```
Space at beginning of line

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_mattermost.md:
````diff
@@ -0,0 +1,152 @@
+```
+
+#### Export access-plugin Certificate
+Teleport Plugin uses the `access-plugin`role and user to peform the approval. We 
````
`access-plugin` role and user (spacing)

`peform` -> `perform`

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_mattermost.md:
```diff
@@ -0,0 +1,152 @@
+
+The above sequence should result in three PEM encoded files being generated: auth.crt,
+ auth.key, and auth.cas (certificate, private key, and CA certs respectively).  We'll
+  reference these later in the bot config, and move them to an appropriate directory.
```
Space at beginning of line

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_mattermost.md:
```diff
@@ -0,0 +1,152 @@
+
+# Checkout teleport-plugins
+git clone git@github.com:gravitational/teleport-plugins.git
+cd teleport-plugins
+
+cd access/mattermost
```
Combine these `cd` commands into `cd teleport-plugins/access/mattermost`

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_mattermost.md:
````diff
@@ -0,0 +1,152 @@
+running the following command: 
+
+```
+teleport-mattermost configure > /etc/teleport-mattermost.yml
````
I think the path is `/etc/teleport-mattermost.toml` if it's a TOML file.

Also you likely won't be able to write to `/etc/teleport` if you're not `root`, so better to write the file to the local directory and then copy it.

```bash
$ teleport-mattermost configure > teleport-mattermost.toml
$ sudo mv teleport-mattermost.toml /etc
```

benarent replied:
Thanks Gus, I've updated the rest of the examples.

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_pagerduty.md:
```diff
@@ -0,0 +1,124 @@
+First off, using an existing Teleport Cluster, we are going to create a new Teleport User and Role to access Teleport.
+
+#### Create User and Role for access. 
+Log into Teleport Authent Server, this is where you normally run `tctl`. Don't change the username and the role name, it should be `access-plugin` for the plugin to work correctly.
```
`the Teleport Auth server`

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_pagerduty.md:
````diff
@@ -0,0 +1,124 @@
+```
+
+#### Export access-plugin Certificate
+Teleport Plugin uses the `access-plugin`role and user to perform the approval. We export the identify files, using [`tctl auth sign`](https://gravitational.com/teleport/docs/cli-docs/#tctl-auth-sign).
````
`access-plugin` role and user (spacing)

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_pagerduty.md:
```diff
@@ -0,0 +1,124 @@
+make access-pagerduty
+
+# Configure the plugin
+./access/pagerduty/build/teleport-pagerduty configure > teleport-pagertudy.toml
```
`pagertudy` -> `pagerduty`

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_pagerduty.md:
````diff
@@ -0,0 +1,124 @@
+Teleport Pagerduty plugin has its own configuration file in TOML format. Before starting the plugin for the first time, you'll need to generate and edit that config file. 
+
+```bash
+teleport-pagerduty configure > /etc/teleport-pagerduty.toml
````
Same thing with not being able to write directly to `/etc/`

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_pagerduty.md:
````diff
@@ -0,0 +1,124 @@
+```
+
+#### Editing the config file
+Afger generating the config, edit it as follows: 
````
`Afger` -> `After`

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_slack.md:
```diff
@@ -0,0 +1,223 @@
+## Teleport Plugins Setup Quickstart
+
+If you're using Slack, you can be notified of [new teleport permission requests](https://gravitational.com/teleport/docs/cli-docs/#tctl-request-ls), approve or deny them on Slack with Teleport Slack Plugin. This guide covers it's setup.
```
`new teleport permission requests` -> `new Teleport permission requests`

`covers it's setup` -> `covers its setup`

benarent replied:
I've rewritten these.

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_slack.md:
```diff
@@ -0,0 +1,223 @@
+- Slack Admin Privileges to create an app and install it to your workspace.
+
+### Create an access-plugin role and user within Teleport 
+First off, using an existing
+Teleport Cluster, we are going to create a new Teleport User and Role to access Teleport.
```
Line spacing

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_slack.md:
````diff
@@ -0,0 +1,223 @@
+```
+
+#### Export access-plugin Certificate
+Teleport Plugin uses the `access-plugin`role and user to perform the approval. We export the identify files, using [`tctl auth sign`](https://gravitational.com/teleport/docs/cli-docs/#tctl-auth-sign).
````
`access-plugin` role and user (spacing)

benarent replied:
I've rewritten these.

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_slack.md:
```diff
@@ -0,0 +1,223 @@
+App Name: Teleport
+Development Slack Workspace: Pick the workspace you'd like the requests to show up in.
+
+![Create Slack App](https://p197.p4.n0.cdn.getcloudapp.com/items/llu4EL7e/Image+2020-01-09+at+10.40.39+AM.png?v=d9750e4fdc77901e0c2ffb2dc6040aee)
```
We should host this image locally

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_slack.md:
```diff
@@ -0,0 +1,223 @@
+
+For now, just think of the URL you'll use and set it in the Slack App's settings screen in Features > Interactive Components > Request URL.
+
+![Interactive Componets](https://p197.p4.n0.cdn.getcloudapp.com/items/v1umg7J0/Image+2020-01-09+at+10.48.08+AM.png?v=5e231f17304f506db13c3ccf445b6682)
```
Locally host the image

benarent replied:
Fixed.

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_slack.md:
```diff
@@ -0,0 +1,223 @@
+
+#### Add to Workspace
+
+![OAuth Token](https://p197.p4.n0.cdn.getcloudapp.com/items/E0uEg1ol/Image+2020-01-09+at+11.00.23+AM.png?v=1e28ff5bc4f7e0754acc9c7823f354a3)
```
Locally host the image

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_slack.md:
````diff
@@ -0,0 +1,223 @@
+
+To generate a config file, you can do this: 
+```bash
+teleport-slackbot configure > /etc/teleport-slackbot.toml
````
Can't write directly to `/etc/`

benarent replied:
Yeah, I found that out https://github.com/gravitational/teleport-plugins/issues/90

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_slack.md:
````diff
@@ -0,0 +1,223 @@
+If everything works fine, the log output should look like this:
+
+```bash
+vm0:~/slackbot sudo ./teleport=slackbot start
````
We should be able to run these plugins without needing to be `root` for security

benarent replied:
Created an issue here https://github.com/gravitational/teleport-plugins/issues/93

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_slack.md:
````diff
@@ -0,0 +1,223 @@
+Seeking request approval... (id: 8f77d2d1-2bbf-4031-a300-58926237a807)
+```
+
+You should now see a new request in Teleport, and a message about the request on Slack. You can approve or deny it and `tsh`should login successfully or error out right after you click an action button on Slack.
````
`tsh` should (spacing)

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_slack.md:
````diff
@@ -0,0 +1,223 @@
+### Setup with SystemD
+In production, we recommend starting teleport plugin daemon via an init system like systemd . Here's the recommended Teleport Plugin service unit file for systemd: 
+
+```
+[Unit]
+Description=Teleport Plugin
+After=network.target
+
+[Service]
+Type=simple
+Restart=on-failure
+ExecStart=/usr/local/bin/teleport-slackbot start --config=/etc/teleport-slackbot.toml --pid-file=/var/run/teleport-slackbot.pid
+ExecReload=/bin/kill -HUP $MAINPID
+PIDFile=/var/run/teleport-slackbot.pid
+
+[Install]
+WantedBy=multi-user.target
+```
````
We should provide these `systemd` units for every plugin.

benarent replied:
I've added these

webvictim reviewed (commented):
Sorry, couldn't resist a review...

benarent commented on docs/4.3/enterprise/workflow/index.md:
```diff
@@ -0,0 +1,132 @@
+  options:
+    # ...
+    # Only allows the contractor to use this role for 1hr from time of request.
+    options.max_session_ttl: 1hr
```
Good catch, I've fixed this.

benarent committed:
Feedback from Gus and cleanup of Jira Cloud

benarent committed:
Extra Fixes

benarent committed:
Port over images and fix Slack Copy

benarent committed:
Update ssh_approval_slack.md

stevenGravy committed:
Minor edits

stevenGravy committed:
added TODO for empty faq

stevenGravy committed:
Minor edit

stevenGravy commented:
I made some minor edits.  One bit to just come back to is the location of the plugins.  I believe this is already on the radar but just mentioning again.

benarent committed:
Update ssh

benarent committed:
Mattermost cleanup

benarent committed:
Add images and fix PD Setup

benarent committed:
PagerDuty Updates

benarent committed:
Improving Slack Guide

benarent committed:
Extra plugin cleanup

benarent committed:
Extra Plugin Cleanup

benarent committed:
Teleport Plugins TOML

benarent committed:
tmp changes to pd

benarent committed:
extra plugin cleanup

benarent committed:
Extra plugin cleanup

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_jira_server.md:
```diff
@@ -1,12 +1,12 @@
 # Teleport Jira Plugin Setup
 
-If you're using Jira Cloud or Jira Server to manage your projects, you can also use it to monitor, approve, deny, or discuss Teleport permission requests. This quickstart will walk you through the setup.
+This guide will talk through how to setup Teleport with Pagerduty.   Teleport ↔ Pagerduty integration  allows you to treat Teleport access and permission requests as Pagerduty incidents — notifying the appropriate team, and approve or deny the requests via Pagerduty special action.
```
Spacing is weird here, lots of double/triple spaces. Also the `↔` character might not exist in some fonts.

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_jira_server.md:
```diff
@@ -1,12 +1,12 @@
 
-For the purpose of this quickstart, we assume you've already setup an [Enterprise Teleport Cluster](https://gravitational.com/teleport/docs/enterprise/quickstart-enterprise/). 
+!!! warning
+    The Approval Workflow only works with Teleport Enterprise as it's requires several roles.
```
`as it requires`

benarent committed:
Extra install instructions.

benarent committed:
More explictic role/username for plugin

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_jira_server.md:
```diff
@@ -49,14 +49,15 @@ $ tctl create -f rscs.yaml
 
-_Note: by default, tctl auth sign produces certificates with a relatively short lifetime. For production deployments, the --ttl flag can be used to ensure a more practical certificate lifetime. --ttl=8760h exports a 1 year token_
+!!! note "Certificate Lifetime"
+     By default, tctl auth sign produces certificates with a relatively short lifetime. For production deployments, the `--ttl` flag can be used to ensure a more practical certificate lifetime. `--ttl=8760h` exports a 1 year token
```
tctl auth sign -> `tctl auth sign`

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_jira_server.md:
```diff
@@ -104,7 +105,18 @@ In the webhook settings page, make sure that the webhook will only send Issue Up
 
-To start using Teleport Plugins, you will need the teleport-jirabot executable.  See the [README](README.md) for building the teleport-jiraport executable in the Setup section.  Place the executable in the appropriate /usr/bin or /usr/local/bin on the server installation.
+
+Where? .. Binary
```
?

benarent replied:
ha, it's updated now :-)

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_mattermost.md:
```diff
@@ -2,8 +2,10 @@
 
-## Setup
+!!! warning
+    The Approval Workflow only works with Teleport Enterprise as it's requires several roles.
```
`as it requires`

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_pagerduty.md:
```diff
@@ -1,11 +1,16 @@
 # Teleport Pagerduty Plugin Setup
 
-This guide will talk through how to setup Teleport with Mattermost.   Teleport ↔  Pagerduty integration  allows you to treat Teleport access and permission requests as Pagerduty incidents — notifying the appropriate team, and approve or deny the requests via Pagerduty special action.
+This guide will talk through how to setup Teleport with Pagerduty.   Teleport ↔ Pagerduty integration  allows you to treat Teleport access and permission requests as Pagerduty incidents — notifying the appropriate team, and approve or deny the requests via Pagerduty special action.
```
weird spacing

webvictim commented on docs/4.3/enterprise/workflow/ssh_approval_pagerduty.md:
```diff
@@ -1,11 +1,16 @@
-## Prerequisites
-This guide assumes you have
+!!! warning
+    The Approval Workflow only works with Teleport Enterprise as it's requires several roles.
```
`as it requires`

benarent committed:
Line spacing cleanup

webvictim reviewed (commented):
quick review

webvictim commented:
Yeah, sorry, I got the email and was a bit overzealous :D

benarent committed:
Extra suggestions from Gus

benarent committed:
Revert "Extra suggestions from Gus"

This reverts commit d3943bbde892801c60be7d0de333b9d36b7e7147.

benarent committed:
Accept all incomming changes.

benarent committed:
Accept all changes

benarent commented:
I seem to be running into a bunch of issues trying to get this into a clean state to merge into 4.3.

@awly The docs-whitespace isn't working for me, I'm getting a bunch ( 169 files, `.md-E`) when I run this. 

```
➜  teleport git:(docs/4.3/benarent/plugins) make docs-fix-whitespace
find: illegal option -- n
usage: find [-H | -L | -P] [-EXdsx] [-f path] path ... [expression]
       find [-H | -L | -P] [-EXdsx] -f path [path ...] [expression]
find docs/ -type f -name '*.md' | xargs sed -i -E 's/\s+$//g'
```

benarent committed:
Cleanup Whitespace

benarent committed:
Extra PR Cleanup

benarent committed:
Fix PAM

benarent committed:
Other cleanup

benarent commented:
Still a WIP but the diff should be much easier to review now. PTAL @webvictim There is also a bunch of duplicate content, I've tried to keep things consistent but not worried about minor difference as long as the instructions are clear.

benarent committed:
Multiple fixes based on PR Feedback

benarent committed:
Fix tctl auth sign

webvictim commented:
> @awly The docs-whitespace isn't working for me, I'm getting a bunch ( 169 files, `.md-E`) when I run this.

It's because you're running on a Mac and GNU `find` is different to whatever MacOS uses.

Commit:
Run docs-fix-whitespace make rule in a busybox container

There are annoying incompatibilities between Linux and MacOS versions of
`find` and `sed`. Instead of fighting them, run in a busybox container
to have predictable behavior.

webvictim committed:
Fix Makefile syntax for Mac

webvictim commented:
@benarent I committed a fix for the `Makefile` on Mac, `git pull` this branch again and it should work.

@awly Turned out to be this: https://github.com/gravitational/teleport/pull/3699/commits/7b20fa434c55e428cdc27585a7e7a79e8aa85c83

benarent commented:
@webvictim odd, that is still giving me these extra -E files 
![image](https://user-images.githubusercontent.com/559288/83073557-7d9e9380-a025-11ea-9cf1-c59d76c7b19d.png)

webvictim commented:
That’s probably what https://github.com/gravitational/teleport/pull/3775 is intended to fix. That PR doesn’t have my fix in though (I committed to this branch rather than 4.3-base) which is why it’s still giving you the `find` error.

benarent committed:
Compress images, add 'gifs' and update slack

benarent committed:
Update to reflect new  slack vs slackbot name

webvictim commented:
@benarent I merged Andrew's fix PR and cherry-picked it here, do a `git pull` and try `make docs-fix-whitespace` again?

benarent committed:
the white walker of whitespace

benarent commented:
Merging this into the base fork, as to address meta build issues in one PR.
